Product Collection - Add Editor UI for missing product reference

imanish003 opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #44878 

This PR addresses the issue of handling missing product reference. For example, if merchant selects a product in the Product Collection block, but then deletes that product, we need to inform the merchant that the selected product is no longer available and provide a way to select a new product.
![image](https://github.com/user-attachments/assets/0210192f-d5bf-4d96-af34-3aae9c75d505)

Key changes:
1. Converted `getProductCollectionUIStateInEditor` to a hook `useProductCollectionUIState` to handle API calls for checking deleted product references.
2. Added a spinner to indicate loading state while checking for deleted product references.
3. Introduced a new UI state `DELETED_PRODUCT_REFERENCE` to handle the missing product state.
4. Updated the existing `ProductPicker` component to handle the new UI state, providing a seamless experience for selecting a new product when the previous reference is missing.

These changes ensure that users can smoothly recover from situations where a selected product has been deleted, maintaining the integrity of the Product Collection block.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

**Pre-requisite:** Install this plugin - 
[register-product-collection-tester.zip](https://github.com/user-attachments/files/16362323/register-product-collection-tester.zip)

#### Case 1: Adding "My Custom Collection - Product Context" to a place where product context doesn't exist

1. Create a new post and add `Product Collection` block.
2. Choose "My Custom Collection - Product Context" block.
3. Choose any product from Editor product picker.
4. Now delete the selected product from: Admin > Products > All Products.
	- _Tip:_ If you know how to manually edit block attributes using Code Editor then you can switch to Code Editor, update the `productReference` attribute with any random product ID. This way, you are manually providing product ID that doesn't exist. You must save the post after updating `productReference` attribute.
5. Once you have deleted the product, refresh the page. 
	- Verify that, you can see following UI with message "Previously selected product is no longer available.". This is the missing product state.
![image](https://github.com/user-attachments/assets/0210192f-d5bf-4d96-af34-3aae9c75d505)
	
6. Now select any other product from the product picker.
	- Verify that, product collection renders with products.
	- Newly selected product is available in "Linked Product" control in inspector controls. (See screenshot below)
![image](https://github.com/user-attachments/assets/f928177f-0a1e-4a0c-a622-36ca9485999b)

7. Publish the post and verify that Frontend is working as expected.

#### Case 2: Adding "My Custom Collection - Product Context" on Single Product template

1. Go to Single Product template and add Product Collection block.
2. Choose "My Custom Collection - Product Context" block.
3. Verify that product picker is not shown in Inspector control.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Product Collection: Added Editor UI for missing product reference

</details>

This new UI state will allow merchants to select a new product.

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

imanish003 committed:
Initial implementation of the missing product state

- Changed `getProductCollectionUIStateInEditor` to a hook `useProductCollectionUIState`.
	- As we added logic to check if selected product reference is deleted which require making an API call. Therefore, I decided to use a hook.
- While making an API call to check if product reference is deleted, I decided to show spinner in the block.
- Introduced new UI state `DELETED_PRODUCT_REFERENCE` to handle the missing product state.
- Updated existing `ProductPicker` component to handle the new UI state.
	- It's better to use existing component for the missing product state, as it already has all the required UI.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

imanish003 committed:
Use getEntityRecord to check if product exists and other improvements

imanish003 committed:
Merge branch 'add/44878-product-collection-handling-missing-product-context' of https://github.com/woocommerce/woocommerce into add/44878-product-collection-handling-missing-product-context

imanish003 committed:
Remove console log

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/ProductPicker.tsx:
**Question:** What should happen when a selected product is deleted but the merchant has not selected a new product yet? 
- On Frontend, should we render nothing until correct product reference is provided?

WunderBart replied:
Can we display elements that are visible only to merchants? 🤔

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/ProductPicker.tsx:
**Question:** I am wondering if it's worth implementing "Unselected state" in the first iteration. IMO It's not a must. What do you think?

For context, here is how it looks like:
![image](https://github.com/user-attachments/assets/57d5813a-db44-45fd-8586-dbc495e73044)

WunderBart replied:
I don't think it's necessary for the mvp. It sounds like something that could be useful across blocks, so I'd move the implementation to its own PR. What do you think?

imanish003 committed:
Add E2E tests for deleted product reference in Product Collection block

This commit introduces new E2E tests to verify the behavior of the Product
Collection block when dealing with deleted product references. The changes
include:

1. New test suite in register-product-collection-tester.block_theme.spec.ts
2. Modification to product-collection.page.ts to support custom product selection
3. Minor update to utils.tsx to handle trashed products

These tests ensure that the Product Collection block correctly handles
scenarios where referenced products are deleted, trashed, or restored,
improving the overall reliability of the feature.

imanish003 commented:
I have marked this PR as ready for review after adding E2E tests.

WunderBart commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx:
```diff
@@ -193,69 +199,121 @@ export const getUsesReferencePreviewMessage = (
+			 */
+			const isArchiveLocationWithTermId =
+				location.type === LocationType.Archive &&
+				( location.sourceData?.termId ?? null ) !== null;
```
nit: I think this is the same as
```suggestion
				location.sourceData?.termId != null;
```

imanish003 replied:
Right, But there is ESList rule because of which I can't use it 🤔  Therefore I went with current one.
![image](https://github.com/user-attachments/assets/d7825967-ab93-42ba-8c64-21ffbc99da2e)

WunderBart replied:
Ah, `eqeqeq` (laughs in eslint)! 😅

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/product-collection/register-product-collection-tester.block_theme.spec.ts:
```diff
@@ -357,3 +357,111 @@ test.describe( 'Product Collection registration', () => {
+			},
+		} );
+
+		// Add "Test Product" only if it doesn't exist
+		if ( products.length === 0 ) {
+			const newProduct = await requestUtils.rest( {
+				method: 'POST',
+				path: 'wc/v3/products',
+				data: {
+					name: 'A Test Product',
+					price: 10,
+				},
+			} );
+			testProductId = newProduct.id;
+		} else {
+			testProductId = products[ 0 ].id;
+		}
+	} );
```
Hmmm, this should not be necessary, as we always start each test with the same DB state. Is it somehow conditionally added with the plugin?

imanish003 replied:
Aha! Good catch. I have made the changes in [6c8ab82](https://github.com/woocommerce/woocommerce/pull/51114/commits/6c8ab8266ccb6ec60cafff2bdad8da2c3e71b7d7)

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/product-collection/register-product-collection-tester.block_theme.spec.ts:
```diff
@@ -357,3 +357,111 @@ test.describe( 'Product Collection registration', () => {
+
+		await editor.saveDraft();
+
+		// Delete the product
+		if ( testProductId ) {
+			await requestUtils.rest( {
+				method: 'DELETE',
+				path: `wc/v3/products/${ testProductId }`,
+			} );
+		}
```
Same here (and below), we shouldn't need to check if the product exists.

imanish003 replied:
Aha! Good catch. I have made the changes in [6c8ab82](https://github.com/woocommerce/woocommerce/pull/51114/commits/6c8ab8266ccb6ec60cafff2bdad8da2c3e71b7d7)

imanish003 committed:
Simplify product creation in Product Collection block test

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/product-collection/register-product-collection-tester.block_theme.spec.ts:
```diff
@@ -357,3 +357,92 @@ test.describe( 'Product Collection registration', () => {
 	} );
 } );
+
+test.describe( 'Deleted product reference in Product Collection block', () => {
```
Let's keep one `describe` per block for good practice. If we need more tests for this context, we should move them to a separate file.

imanish003 replied:
Good point! Made the changes in [543f6d9](https://github.com/woocommerce/woocommerce/pull/51114/commits/543f6d943af7570cd5a1d3bebd1caf1c766ee655)

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/product-collection/register-product-collection-tester.block_theme.spec.ts:
```diff
@@ -357,3 +357,92 @@ test.describe( 'Product Collection registration', () => {
 } );
+
+test.describe( 'Deleted product reference in Product Collection block', () => {
+	let testProductId: number | null = null;
+
+	test.beforeEach( async ( { requestUtils } ) => {
+		// Activate plugin
+		await requestUtils.activatePlugin(
+			'register-product-collection-tester'
+		);
+
+		// Add a new test product to the database
+		const newProduct = await requestUtils.rest( {
+			method: 'POST',
+			path: 'wc/v3/products',
+			data: {
+				name: 'A Test Product',
+				price: 10,
+			},
+		} );
+		testProductId = newProduct.id;
+	} );
+
+	test( 'Product picker should be shown when product reference is deleted', async ( {
+		pageObject,
+		admin,
+		editor,
+		requestUtils,
+	} ) => {
+		await admin.createNewPost();
```
Let's drop the `beforeEach` hook if there's just one test.

imanish003 replied:
Makes sense! Made the changes in [543f6d9](https://github.com/woocommerce/woocommerce/pull/51114/commits/543f6d943af7570cd5a1d3bebd1caf1c766ee655)

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/index.tsx:
```diff
@@ -31,14 +32,25 @@ const Edit = ( props: ProductCollectionEditComponentProps ) => {
-			usesReference: props.usesReference,
+			attributes,
+			hasInnerBlocks,
+			...( props.usesReference && {
+				usesReference: props.usesReference,
+			} ),
```
What's the benefit of not passing `usesReference` if it's not available? Because I think `usesReference: props.usesReference,` is cleaner and more explicit, even if sometimes value will be `undefined`.

WunderBart replied:
I bet it's just a TS thing 😅

imanish003 replied:
Yeah, It was TS thing but I was able to simplify it in [7380aa8](https://github.com/woocommerce/woocommerce/pull/51114/commits/7380aa818e946ca0f802ee1d799a92c4dc4b94b9) 😀

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/index.tsx:
```diff
@@ -73,6 +86,10 @@ const Edit = ( props: ProductCollectionEditComponentProps ) => {
 				isUsingReferencePreviewMode={ isUsingReferencePreviewMode }
 				location={ location }
 				usesReference={ props.usesReference }
+				isDeletedProductReference={
+					productCollectionUIStateInEditor ===
+					ProductCollectionUIStatesInEditor.DELETED_PRODUCT_REFERENCE
+				}
```
I understand we need it to differentiate `ProductPicker` but I'm not a fan of this solution in which we pass the prop to all the other components as well.

One option to overcome this is to extract `sharedProps` to an object and in the switch above, return specific component, rather than assign `Component = ...`.

In other words, it would be:

```
case ProductCollectionUIStatesInEditor.COLLECTION_PICKER:
	return <ProductCollectionPlaceholder {...sharedProps} />;
case ProductCollectionUIStatesInEditor.PRODUCT_REFERENCE_PICKER:
    return <ProductPicker {...sharedProps} />;
case ProductCollectionUIStatesInEditor.DELETED_PRODUCT_REFERENCE:
	return <ProductPicker {...sharedProps}  isDeletedProductReference />;
...
```

I think it would be cleaner and simpler, wdyt? 😅

imanish003 replied:
Thank you for the suggestion, Karol! TBH, I have been thinking about this but was avoiding it because I thought it would require a good amount of refactoring 😅 But after reading this feedback, I decided to do it and have made the changes 🙇‍♂️

kmanijak reviewed (commented):
I tested the changes and works just fine! 🚀 

I left some comments regarding the code style. Let me know what you think of them. 🙌

imanish003 committed:
Refactor E2E test for delete product reference picker

1. Removing the unnecessary `test.describe` block for "Deleted product reference"
2. Eliminating the `beforeEach` hook that was creating a test product
3. Integrating the test product creation directly into the main test

This change simplifies the test structure and improves readability while
maintaining the same test coverage for the Product Collection block's
behavior when dealing with deleted or unavailable products.

imanish003 committed:
Simplify logic for product collection UI state

This commit simplifies the handling of the `usesReference` prop in the
Product Collection block:

1. In `edit/index.tsx`, directly pass `usesReference` to
   `useProductCollectionUIState` hook without conditional spreading.

2. In `utils.tsx`, update the type definition of `usesReference` in the
   `useProductCollectionUIState` hook to explicitly include `undefined`.

imanish003 committed:
Refactor Product Collection block to improve prop passing

- Introduce ProductCollectionContentProps type for better prop management
- Refactor Edit component to use a renderComponent function
- Update component prop types to use more specific props
- Remove unnecessary props from ProductCollectionEditComponentProps
- Simplify component rendering logic in Edit component
- Adjust ProductPicker to receive only required props
- Update imports and prop types in various files to use new types

This refactoring improves code organization and reduces prop drilling by
only passing necessary props to each component. It enhances maintainability
and readability of the Product Collection block and related components.

imanish003 committed:
Refactor Product Collection block editor UI state handling

This commit simplifies the rendering logic in the Product Collection block's
Edit component. It removes a redundant case for the VALID state, as both
VALID and VALID_WITH_PREVIEW states now render the same ProductCollectionContent
component. This change improves code maintainability and reduces duplication.

imanish003 commented:
Hi @WunderBart @kmanijak, I believe I have addressed all the feedback. Can you please review the PR again?

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/index.tsx:
```diff
@@ -31,49 +33,59 @@ const Edit = ( props: ProductCollectionEditComponentProps ) => {
+		switch ( productCollectionUIStateInEditor ) {
+			case ProductCollectionUIStatesInEditor.COLLECTION_PICKER:
+				return <ProductCollectionPlaceholder { ...props } />;
+			case ProductCollectionUIStatesInEditor.PRODUCT_REFERENCE_PICKER:
+			case ProductCollectionUIStatesInEditor.DELETED_PRODUCT_REFERENCE:
+				return (
+					<ProductPicker
+						{ ...props }
+						isDeletedProductReference={ true }
+					/>
+				);
```
PRODUCT_REFERENCE_PICKER case requires a separate return without `isDeletedProductReference`. Because in the current shape we display "Previously selected product is no longer available." message even if adding new collection with no previous reference.

```suggestion
			case ProductCollectionUIStatesInEditor.PRODUCT_REFERENCE_PICKER:
			    return <ProductPicker { ...props } />;
			case ProductCollectionUIStatesInEditor.DELETED_PRODUCT_REFERENCE:
				return (
					<ProductPicker
						{ ...props }
						isDeletedProductReference={ true }
					/>
				);
```

imanish003 replied:
Aha! Good catch. Fixed it in 9f04f11

kmanijak reviewed (commented):
Thanks for addressing my feedback and refactoring `renderComponent`! I left one important comment but requiring only a small adjustment. Once fixed, I'll make a final round of testing 🙌

imanish003 committed:
Fix: isDeletedProductReference is not set correctly

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/product-collection/register-product-collection-tester.block_theme.spec.ts:
```diff
@@ -356,4 +356,83 @@ test.describe( 'Product Collection registration', () => {
+		} );
+
+		// Product picker should be shown in Editor
+		await admin.page.reload();
```
Let's use the `page` fixture directly here and below. Another fixture's `page` member is not always available (private), so it might be confusing DX-wise and prone to becoming unavailable. 

```suggestion
		await page.reload();
```

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/index.tsx:
```diff
@@ -31,49 +33,62 @@ const Edit = ( props: ProductCollectionEditComponentProps ) => {
+		switch ( productCollectionUIStateInEditor ) {
+			case ProductCollectionUIStatesInEditor.COLLECTION_PICKER:
+				return <ProductCollectionPlaceholder { ...props } />;
+			case ProductCollectionUIStatesInEditor.PRODUCT_REFERENCE_PICKER:
+			case ProductCollectionUIStatesInEditor.DELETED_PRODUCT_REFERENCE:
+				return (
+					<ProductPicker
+						{ ...props }
+						isDeletedProductReference={
+							productCollectionUIStateInEditor ===
+							ProductCollectionUIStatesInEditor.DELETED_PRODUCT_REFERENCE
+						}
+					/>
+				);
```
Super nit: I'm not a fan the situation in which we have two cases pointing to single return but then there's a condition inside taking into account the `case` condition 😅 

So we have:
```
case A:
case B:
    return <ProductPicker prop={ condition === B }
```

while I think we could be more explicit:

```
case A:
   return <ProductPicker prop={ false }
case B:
    return <ProductPicker prop={ true }
```

But it's just some sort of preference hence don't feel it has to be addressed.

imanish003 replied:
I don't have a strong opinion so made the changes in 6be9f48 🤝

kmanijak reviewed (approved):
Tested and works just fine! Left one nitpick comment but approving anyway! Great job as always 👏

imanish003 committed:
Use "page.reload" instead of "admin.page.reload"

imanish003 committed:
Separate PRODUCT_REFERENCE_PICKER and DELETED_PRODUCT_REFERENCE cases

This improves readability and maintainability of the switch statement.
